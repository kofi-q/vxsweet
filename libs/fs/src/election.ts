import { type Result, err } from '@vx/libs/basics/result';
import { type ElectionDefinition } from '@vx/libs/types/elections';
import { safeParseElectionDefinition } from '@vx/libs/types/election-parsing';
import { ZodError } from 'zod';
import { type ReadFileError, readFile } from './read_file';

/**
 * Possible errors that can occur when reading an election.
 */
export type ReadElectionError =
  | { type: 'ReadFileError'; error: ReadFileError }
  | { type: 'ParseError'; error: ZodError | SyntaxError };

/**
 * Maximum size of an election file (30 MB).
 */
const MAX_ELECTION_SIZE = 30 * 1024 * 1024;

/**
 * Reads an election from a file path.
 */
export async function readElection(
  electionPath: string
): Promise<Result<ElectionDefinition, ReadElectionError>> {
  const readFileResult = await readFile(electionPath, {
    maxSize: MAX_ELECTION_SIZE,
    encoding: 'utf-8',
  });

  if (readFileResult.isErr()) {
    return err({ type: 'ReadFileError', error: readFileResult.err() });
  }

  const parseResult = safeParseElectionDefinition(readFileResult.ok());

  if (parseResult.isErr()) {
    return err({ type: 'ParseError', error: parseResult.err() });
  }

  return parseResult;
}
