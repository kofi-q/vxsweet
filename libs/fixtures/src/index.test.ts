import * as fixtures from '.';
import { election as electionGeneral } from '@vx/libs/fixtures/src/data/electionGeneral/election.json';

test('has various election definitions', () => {
  expect(
    Object.entries(fixtures)
      .filter(([, value]) => typeof value !== 'function')
      .map(([key]) => key)
      .sort()
  ).toMatchSnapshot();
});

test('asElectionDefinition', () => {
  expect(fixtures.asElectionDefinition(electionGeneral)).toStrictEqual(
    expect.objectContaining({
      election: electionGeneral,
      electionData: expect.any(String),
      ballotHash: expect.any(String),
    })
  );
});
