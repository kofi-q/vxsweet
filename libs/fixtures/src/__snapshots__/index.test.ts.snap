// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has various election definitions 1`] = `[]`;
